### ISSUE TITLE
Further Cache Management

### ISSUE THREAD

**Issue by eddableheath** [alan-turing-institute/ARC-LoCoMoSeT#68]
Following on from #28 (which is now solely concerned with caching location), we should be tracking how much storage the caching is using and either:

- Implementing automatic clearing of the caches when the datasets, models and model checkpoints are no longer needed.
- Developing good practice to delete these manually when necessary.,

**Comment by jack89roberts** [comment]
Datasets: https://huggingface.co/docs/datasets/cache
Models: https://huggingface.co/docs/huggingface_hub/guides/manage-cache

For datasets, preprocessed images (i.e. after being passed through the `ImageProcessor`, I think not for full forward passes for how we've currently implemented it based on the inference time in the jobs) get cached for each model, so the dataset cache can get very large very quickly. E.g. the Oxford Pets dataset is ~250 MB but the cache on Baskerville was ~150 GB (now deleted).

From a quick look at the docs disabling the cache would still save files to a temporary directory, and those files would then not be shared between the same model, e.g. if we had 5 jobs running for the same model/dataset at the same time I think wherever the temporary dirs are would need enough space to store 5 duplicates of the cache files.

**Comment by jack89roberts** [comment]
Update after some trials on `wikiart`:

- Caching to the project directory can fill up our (1 TB) quota if we're running jobs for multiple models at the same time. Also, if jobs don't finish/complete nicely they can leave behind some huge `tmp...` files in the cache directory.
- You can disable caching in HuggingFace `datasets` (see [here](https://huggingface.co/docs/datasets/cache)), but then it caches to a temporary directory instead. I think Baskerville jobs get a local `/tmp` with a 100 GB quota which didn't seem to be enough. Or might have been some other issue with where HuggingFace was trying to put the tmp files. Maybe could see if the temporary directory can be somewhere else, e.g. [scratch-global](https://docs.baskerville.ac.uk/storage/#temporary-scratch-storage)?
- You can tell datasets not to cache stuff to disk but rather keep everything in RAM instead (also see [here](https://huggingface.co/docs/datasets/cache)). I got that to work, you get 100 GB RAM per GPU requested on Baskerville. But, for a few models, I had to request 2 or 3 GPUs to get the dataset to fit in RAM.

For `rvl_cdip` just chucking more resources at the problem probably won't work I suspect, but I have a trial job queued to see. The main longer-term solutions I see are either:

1) Find a way/place for each job to have much more temporary storage available.
2) Run fewer jobs in parallel and tidy up the project directory cache properly between them.
3) (Probably the proper solution) Look at implementing a more appropriate data loader that would get around this (i.e. that caches less to disk or RAM). Or maybe using streaming datasets would get around this (depending on their caching behaviour - it may just delay the issue).

**Comment by jack89roberts** [comment]
> * You can disable caching in HuggingFace `datasets` (see [here](https://huggingface.co/docs/datasets/cache)), but then it caches to a temporary directory instead. I think Baskerville jobs get a local `/tmp` with a 100 GB quota which didn't seem to be enough. Or might have been some other issue with where HuggingFace was trying to put the tmp files. Maybe could see if the temporary directory can be somewhere else, e.g. [scratch-global](https://docs.baskerville.ac.uk/storage/#temporary-scratch-storage)?

I think setting `TMPDIR`, `TEMP` or `TMP` environment variables should change where `tempfile` puts temporary files per default (and `datasets` is using `tempfile` to specify the temp dir location). E.g see docs of https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp

Edit: See #93

**Comment by jack89roberts** [comment]
### `datasets` args that control caching behaviour:

- `keep_in_memory=False or True` in `dataset.map`: Keep the dataset in RAM rather than caching to disk (set True if `preprocess_cache: ram` in the `caches` config). However RAM usage seems to increase during preprocessing even if this is False, there are some issues around this e.g. [here](https://github.com/huggingface/datasets/issues/4883)
- `datasets.disable_caching()`: Cache files to a temporary directory that's deleted after the training job has finished (set if `preprocess_cache: tmp` in the `caches` config, and location of temp dir set by `tmp_dir`)
- `writer_batch_size` in `dataset.map` (defaults to 1000): Controls how many samples will be stored in RAM before caching them to disk (smaller values mean slower preprocessing but less RAM usage, set by `writer_batch_size` in `caches` config)

### Baskerville args:

- By default you get 128 GB of RAM per GPU (512 GB RAM per full node). So requesting more GPUs is a way to avoid out of memory errors (if `keep_in_memory=True`). Slurm has a `mem-per-gpu` arg that can be set which I've not tried but could be a way of just asking for more RAM without more GPUs (if Baskerville lets you configure that).
- Disk quotas:
  - Project directory: 1 TB
  - Per-job `/tmp` (i.e. `/scratch-local`): 500 GB (increased from 100 GB after our support query)
  - `/scratch-global`: 193 TB shared between all Baskerville users (may be asked to delete files if overused and shouldn't be used for persistent storage).
- Caching in RAM or to `/tmp` (`/scratch-local`) should be preferred to caching to the project directory or `/scratch-global`, where possible. It will be faster and reduce network load.

### Reference values:

- `microsoft/beit-large-patch16-512` on `rvl_cdip` looks like (job is running now) it will end up caching at least 1 TB, which is huge but does make sense given the dataset size and model input resolution:
  - 512 x 512 x 3 (RGB) image resolution ~= 786k pixel values
  - 32-bit precision * 786k values ~= **3.1 MB per image**
  - 400k images (train + test + val) * 3.1 MB per image ~= **1.2 TB for the whole dataset**
- Alternatively, a model with an input resolution of 224 x 224 x 3 would use **235 GB (0.6 MB per image)** - I believe this is the smallest input resolution out of the models we're using (but also the most common)
- Presumably the Intel 8-bit model with 224 x 224 x 3 could use up to 4x less (~60 GB)
- The above is for the preprocessed images only and doesn't account for anything else that might be stored in the cache (e.g. non-pixel tokens, class labels and other metadata)

**Comment by jack89roberts** [comment]
With the caveats above, this is the minimum cache size (in GB) required to store the preprocessing data for the datasets we've tried so far at different input resolutions (if using the whole dataset):

Dataset | No. Images | 224 | 256 | 384 | 512
-- | -- | -- | -- | -- | --
pcuenq/oxford-pets | 7,390 | 4 | 6 | 13 | 23
huggan/wikiart | 81,444 | 48 | 63 | 141 | 250
rvl-cdip | 400,000 | 235 | 307 | 691 | 1229

Rough hardware requirements:

- Oxford Pets: Single GPU and caching to RAM should work for all models
- Wikiart: Single GPU and cache to `/tmp` should work for all models (or 2/3 GPUs for 384/512 input resolutions if wanting to cache to RAM)
- RVL CDIP: Single GPU and caching to `/scratch-global` should work for all models (could cache to `/tmp` for 224/256 input resolutions or to RAM for those resolutions if requesting 2/3 GPUs)

We might want to settle on multi-GPUs for some datasets for other reasons, i.e. to make training faster. Or, more sensibly, just not training on the whole dataset (especially for RVL CDIP).

**Reminder:** Cache sizes above only apply to training, nothing (should be) cached for metrics (apart from the original dataset).

**Comment by jack89roberts** [comment]
TODO:
- Check the back of the envelope calculations above work in practice (i.e. will training of all models run within the limits specified above)
- Choice of `writer_batch_size` and impact of that on runtimes
- Impact of other caching choices on runtimes
- Consider how to factor in any differences in metrics vs. training hardware in time taken analysis
- Consider how to set appropriate walltime for jobs across datasets and models
- Update configs accordingly
